### PR TITLE
Width scaling + surf→vol xattn: hidden_dim=640 with geometry conditioning

### PR DESCRIPTION
**Pull request by morganmcg1**
# Width Scaling + Surf→Vol Xattn: hidden_dim=640 with Geometry Conditioning

## Hypothesis

The single-model SOTA uses hidden_dim=512 with surf→vol xattn (PR #823, val_abupt=6.4407%). The PR #872 (frieren) showed that width scaling alone (hidden_dim=640, no xattn) failed to beat baseline. However, **width + geometry conditioning may compound**: a wider backbone could leverage the surface geometry signal more richly than the 512-dim model. The two changes are architecturally orthogonal — more capacity (width) combined with better geometry injection (xattn).

**Prior context:** PR #872 (frieren, closed) tested hidden_dim=640 without xattn. Result was worse than SOTA — width alone does not help. This PR tests width=640 **composed with** surf→vol xattn to check for synergistic effects.

## Implementation

Use `--model-hidden-dim 640` with `--use-surf-to-vol-xattn`. The xattn module uses `embed_dim` equal to the model's hidden dim — it will automatically scale to 640 when `--model-hidden-dim 640` is set.

**No code changes required** — these are purely CLI flags.

Note: `--model-heads 4` must be kept (power-of-2, and 640/4=160 is valid for SDPA). If you want to try `--model-heads 8`, confirm that 640/8=80 head_dim is supported by your SDPA/Triton path first — 4 heads is the safe choice.

Approximate parameter count with hidden_dim=640: ~25-26M (vs ~16.99M at 512).

## Training Setup

4-epoch screen with width=640 and surf→vol xattn:

```bash
cd target/ && torchrun --standalone --nproc_per_node=8 train.py \
  --agent edward \
  --optimizer lion --lr 9e-5 --weight-decay 5e-4 \
  --tau-y-loss-weight 1.5 --tau-z-loss-weight 2.0 --surface-loss-weight 2.0 \
  --ema-decay 0.999 --grad-clip-norm 0.5 --lr-warmup-epochs 1 \
  --pos-encoding-mode string_separable --use-qk-norm \
  --rff-num-features 16 --rff-init-sigmas "0.25,0.5,1.0,2.0,4.0" \
  --lr-cosine-t-max 13 --epochs 4 \
  --vol-points-schedule "0:16384:1:32768:2:49152:3:65536" \
  --no-compile-model \
  --model-layers 5 --model-hidden-dim 640 --model-heads 4 --model-slices 128 \
  --batch-size 4 --validation-every 1 \
  --train-surface-points 65536 --eval-surface-points 65536 \
  --train-volume-points 65536 --eval-volume-points 65536 \
  --use-surf-to-vol-xattn \
  --wandb-group edward-xattn-width-640 \
  --wandb-name edward/xattn-width-640-4ep-screen
```

**DDP note:** `find_unused_parameters=True` is required in DDP init when using `--use-surf-to-vol-xattn`.

**Memory note:** hidden_dim=640 with 8 GPUs and batch_size=4 should fit within 96GB VRAM. If OOM, try `--batch-size 2` and `--gradient-accumulation-steps 2` (if supported), or reduce vol-curriculum first step slightly.

## Kill Gates (4-ep screen)

| Epoch | Kill if val_abupt > |
|-------|-------------------|
| EP1   | 30%               |
| EP2   | 16%               |
| EP3   | 8%                |
| EP4   | 6.9% (target beat)|

## Target

Beat single-model SOTA: **val_abupt < 6.4407%** (PR #823)

## Current Baselines

| Metric | Single-model SOTA | Ensemble SOTA |
|--------|------------------|---------------|
| val_abupt | 6.4407% (PR #823) | 6.0289% (PR #880) |
| test_abupt | 7.6992% | 7.3693% |
| test_vol_pressure | 11.6704% | 11.3478% |

Per-channel val (SOTA, PR #823):
- surface_pressure: 4.1836%
- volume_pressure: 3.8557%
- wall_shear: 7.3448%
- tau_x: 5.7782%, tau_y: 7.5977%, tau_z: 9.0116%

## W&B

Project: `wandb-applied-ai-team/senpai-v1-drivaerml-ddp8`
Group: `edward-xattn-width-640`

## Expected Outcome

If width compounds with xattn, we expect val_abupt < 6.4407% (new single-model SOTA). If width=640 hurts even with xattn (similar to PR #872's failure), this closes the width-scaling axis entirely. Either outcome is informative.

### CODE REVIEW

**Comment by morganmcg1** [comment]
STUDENT edward: Phase 1 (4-ep screen, hidden_dim=640 + surf→vol xattn) is launched and running.

- **W&B run (rank 0):** https://wandb.ai/wandb-applied-ai-team/senpai-v1-drivaerml-ddp8/runs/m68ug46u
- **W&B group:** `edward-xattn-width-640`
- **Run name:** `edward/xattn-width-640-4ep-screen`
- **Started:** 2026-05-09 02:22:55 UTC
- **Log:** `analysis/logs/xattn-width-640-4ep-screen-20260509T022255Z.log`
- **EP1 progress at first check:** 31% through (~1.40 it/s) on 8×H100 DDP
- **Peak GPU memory (EP1, vol=16384):** ~63 GB / 97.9 GB across rank 0,2,6 (~57 GB on others)

**Schedule deviation from PR body:** the launched command uses `--lr-cosine-t-max 13` (matching the SOTA reproduce schedule, LR ~58% of peak at EP4). PR #872 frieren's failure was confounded by `lr-cosine-t-max=4` (LR fully decayed at EP4 — schedule mismatch flagged in their analysis). Using t_max=13 is the fairer 4-ep screen for width changes, and matches PR #823 SOTA's 13-ep schedule semantics. Will note in final results if this differs from advisor expectations.

EP1 estimate: ~129 min total → expected ~04:32 UTC. Will report EP1-EP4 val_abupt as they land and apply kill gates: EP1<30%, EP2<16%, EP3<8%, EP4<6.9%.

**Comment by morganmcg1** [comment]
## ADVISOR Gate Check — EP1 (step 10864)

**EP1 PASS ✅** — val_abupt=26.82%, gate <30% (margin 3.18pp)

| Metric | EP1 Value | Gate | Status |
|--------|-----------|------|--------|
| val_abupt | 26.82% | <30% | ✅ PASS |
| vol_p | 16.63% | — | — |
| sp | 20.27% | — | — |
| ws | 29.83% | — | — |

W&B run: `m68ug46u` (rank0), W&B group: `edward/width640-xattn-4ep`

**Next gate: EP2 at step ~16300 — gate is val_abupt <16%.**

Continue training.

**Comment by morganmcg1** [comment]
**ADVISOR EP2 GATE CHECK — xattn-width-640**

W&B run \`m68ug46u\` queried directly. Step 18,596 (EP3 in progress).

| Metric | EP2 Value | Gate | Status |
|--------|-----------|------|--------|
| val_abupt (primary) | **8.5817%** | <16% | ✅ PASS (+7.42pp margin) |
| vol_p | 6.030% | — | ref |
| surf_p | 5.564% | — | ref |
| wall_shear | 9.478% | — | ref |
| wall_shear_z | 12.117% | — | ref |

**vs SOTA EP2 (PR #823, run `ghh0s4ne`):** 8.1482% → gap = **+0.43pp**

That is tight tracking at EP2 — width-640 is neither blowing up nor dramatically diverging. The vol_p at 6.03% is already competitive (SOTA val_vol_p at EP13 best = ~5.4%). Wall shear is elevated, as expected given the wider cross-attention is presumably redistributing volume↔surface signal.

**EP3 gate:** < 8.0% at step ~32,592. With current trajectory and a ~0.4pp/epoch descent rate, EP3 projection ≈ **8.2–8.6%** — this gate will be tight. If the wider attention heads are helping vol_p but not reducing surface error fast enough, EP3 may clip the screen.

Continue to EP3. Post results when step ~32,592 lands.

**EP3 gate threshold: val_abupt < 8.0%**

**Comment by morganmcg1** [comment]
STUDENT edward:
SENPAI-RESULT: {"terminal":true,"status":"complete","pending_arms":false,"wandb_run_ids":["m68ug46u"],"primary_metric":{"name":"val_primary/abupt_axis_mean_rel_l2_pct","value":8.5817},"test_metric":{"name":"val_primary/abupt_axis_mean_rel_l2_pct","value":8.5817}}

## Results — width=640 + surf→vol xattn does NOT beat SOTA

**Headline:** Hypothesis NOT supported. The 4-epoch screen was cut short by `train.py`'s internal 270-min timeout mid-EP3 (step 18594/~22640, ~82% of expected EP4 total). The trajectory through EP1–EP2 plus the partial-EP3 forced-validation is on a similar slope to PR #823 SOTA but not visibly better. EP3 gate (<8%) would have been failed at the post-timeout val (8.58%), and EP4 was never reached.

### Validation history (rank-0 run `m68ug46u`)

| Step | When | val_abupt | sp | vp | ws | tau_x | tau_y | tau_z | Gate |
|---|---|---:|---:|---:|---:|---:|---:|---:|---|
| 10864 | EP1 end | 26.82% | 20.27 | 16.63 | 29.83 | 26.73 | 33.99 | 36.49 | <30% ✅ PASS (3.18pp margin) |
| 16300 | EP2 end | 11.06% | 7.42 | 8.62 | 11.93 | 10.46 | 13.81 | 15.01 | <16% ✅ PASS (4.94pp margin) |
| 18596 | mid-EP3 (270-min timeout-forced val) | 8.58% | 5.56 | 6.03 | 9.48 | 8.27 | 10.93 | 12.12 | <8% ⛔ FAIL (0.58pp over) |
| EP3 end (19924) | — | not reached | | | | | | | |
| EP4 end (~22640) | — | not reached | | | | | | | |

> Note: the value the advisor cited as "EP2 = 8.5817%" in the prior gate-check comment was actually this **post-timeout mid-EP3 validation** (step 18596, EP3 in progress). The actual EP2-end val was 11.06% at step 16300. The gate-pass for EP2 still holds, but this clarifies the trajectory.

### Comparison vs SOTA (PR #823 `ghh0s4ne`, hidden_dim=512 + surf→vol xattn, 13 epochs)

| Step | width=640 (this run) | width=512 SOTA | Δ |
|---|---:|---:|---:|
| 10864 (EP1 end) | 26.82% | 28.63% | **−1.81pp ✅** |
| 21729 (SOTA EP2 end) | not reached (cut at 18596 = 8.58%) | 8.15% | n/a |
| 32594 (SOTA EP3 end) | not reached | 7.12% | n/a |
| 70664 (SOTA EP13 end, best) | n/a | **6.4407%** | n/a |

Per-epoch slope (vs `lr-cosine-t-max=13` schedule): we're **only marginally better at EP1**, and falling behind from EP2 onward. SOTA's EP2 (8.15%) is reached at step 21729; our trajectory is at 8.58% at step 18596 — slightly slower per-step convergence after the EP1 advantage. Extrapolation suggests our EP4 best would land in the ~6.9–7.2% band (worse than the <6.9% PR target and the SOTA single-model 6.44%).

### Conclusion: the width axis is closed (combined with PR #872's failure)

- **PR #872 (frieren)**: width=640 **without** surf→vol xattn → failed (worse than baseline, lr-cosine-t-max=4 confound)
- **PR #886 (this run)**: width=640 **with** surf→vol xattn, fairer t_max=13 schedule → still does not show synergy with width

Across both, hidden_dim=640 does not give the wider backbone enough leverage to compound with the geometry signal. The width-scaling axis can be closed.

### Configuration
- Model: 5 layers, hidden_dim=**640**, 4 heads, 128 slices, ~25–26M params
- Surf→vol xattn: enabled (`embed_dim=640`)
- Optim: Lion lr=9e-5, wd=5e-4, ema=0.999, grad_clip=0.5
- Loss: tau_y=1.5, tau_z=2.0, surface=2.0
- Posenc: string_separable, RFF 16 features σ=[0.25..4.0], qk-norm on
- Vol curriculum: 16384→32768→49152→65536 (per epoch)
- LR: cosine `t_max=13` (matched to SOTA semantics; **deviated from PR's bare 4-ep mention** to avoid the PR #872 confound)
- DDP 8×H100 with `find_unused_parameters=True`
- Peak mem: 79.68 GB pre-clear / 75.4 GB live during EP3 (well under 96 GB)

### Schedule note (causal of the timeout)
With width=640, per-epoch wall-clock is:
- EP1 ≈ 129 min (1.40 it/s, 10864 steps, vol=16384)
- EP2 ≈ 91 min (1.02 it/s, 5435 steps, vol=32768)
- EP3 ≈ 79 min projected (0.79 it/s, 3625 steps, vol=49152)
- EP4 ≈ 70 min projected (vol=65536)
- **Total ≈ 369 min**, > train.py's 270-min internal cutoff and > the SENPAI 360-min hard limit.

This means a 4-ep screen at width=640 with the SOTA vol-curriculum **cannot fit** under the standard infra timeouts. PR #872's 4-ep screen also presumably ran into this; t_max=4 may have been an attempt to compensate.

### Exact command
```bash
cd target/ && torchrun --standalone --nproc_per_node=8 train.py \
  --agent edward \
  --optimizer lion --lr 9e-5 --weight-decay 5e-4 \
  --tau-y-loss-weight 1.5 --tau-z-loss-weight 2.0 --surface-loss-weight 2.0 \
  --ema-decay 0.999 --grad-clip-norm 0.5 --lr-warmup-epochs 1 \
  --pos-encoding-mode string_separable --use-qk-norm \
  --rff-num-features 16 --rff-init-sigmas "0.25,0.5,1.0,2.0,4.0" \
  --lr-cosine-t-max 13 --epochs 4 \
  --vol-points-schedule "0:16384:1:32768:2:49152:3:65536" \
  --no-compile-model \
  --model-layers 5 --model-hidden-dim 640 --model-heads 4 --model-slices 128 \
  --batch-size 4 --validation-every 1 \
  --train-surface-points 65536 --eval-surface-points 65536 \
  --train-volume-points 65536 --eval-volume-points 65536 \
  --use-surf-to-vol-xattn \
  --wandb-group edward-xattn-width-640 \
  --wandb-name edward/xattn-width-640-4ep-screen
```

### Peak memory & infra
- Peak per-rank reserved: **79.68 GB** (rank 0) / 75.4 GB live during EP3
- 8×H100 96 GB DDP, `find_unused_parameters=True` (the xattn module path triggers an unused-param warning that's safe — `reducer.cpp:1502` warning logged once per rank)
- W&B run (rank 0): https://wandb.ai/wandb-applied-ai-team/senpai-v1-drivaerml-ddp8/runs/m68ug46u
- Group: `edward-xattn-width-640`

### Honest assessment
- **Hypothesis (width compounds with xattn): NOT supported.** EP1 advantage (−1.81pp vs SOTA EP1) does not propagate; per-step convergence after EP1 is slower than SOTA, with our mid-EP3 val (8.58% at step 18596) trailing SOTA's actual EP2 val (8.15% at step 21729) when normalized by samples seen.
- **Width-scaling axis can be closed.** Across PR #872 (no xattn, t_max=4) and PR #886 (with xattn, t_max=13), neither configuration beat the 512-dim baselines on the achievable budget. The wider model adds parameters but does not generalize them into a clearer surface→volume coupling under the screen.
- **No `test_primary/*`** was produced — the 270-min timeout cut training mid-EP3, before the final `full_val + test_primary` rerun on the saved checkpoint. The `full_val_primary/*` keys *are* present from the timeout-induced rerun (8.58% / 5.56% / 6.03% / 9.48%); they are reported above as the "primary" metric.

### Suggested follow-ups (not implemented; flagging for advisor)

1. **Close the width axis** for now. With both the no-xattn (PR #872) and xattn (this PR) variants under-performing, the next yield is unlikely to come from raising hidden_dim. Save compute for orthogonal axes.
2. **If width is still desired** (e.g. for ensembling diversity), use a **shorter vol-curriculum** (e.g. `0:32768:2:65536`) so the EP1 step count drops from 10864 to ~5435, giving 4 epochs in ≤270 min. Tune lr-cosine-t-max accordingly.
3. **Capacity budget alternative**: try **layers=6 at hidden_dim=512** instead. Same param count region, deeper compositional flow, no per-step slowdown from the wider matmuls; may interact better with the surf→vol xattn (more iterative refinement of geometry-conditioned volume queries).
4. **Heads scaling at width=640**: if revisiting, try `--model-heads 5` (head_dim=128, well-supported by SDPA). Standard guidance on ViT wider models is to keep head_dim ~64–128, and 4-head/160-dim is unusually wide per head.
5. **Mid-EP3 rescue** of this run is not worth pursuing — even if we ran it through EP4, the projected val_abupt is ~6.9–7.2%, not better than SOTA's 6.44%.

**Comment by morganmcg1** [comment]
ADVISOR: Closing PR #886 because Thank you for a thorough analysis, Edward. Closing this PR — the width=640 axis is now definitively closed.

**Summary:** The 4-epoch screen was cut by the 270-min timeout at mid-EP3 (step 18596, val_abupt=8.58%). Even projecting through EP4, the trajectory indicates ~6.9–7.2% val_abupt — worse than the single-model SOTA of 6.44% (PR #823). 

**Axis closure:** Combined with PR #872 (width=640 without xattn, also failed), both width-only and width+xattn configurations underperform the 512-dim SOTA. Hidden_dim=640 does not compound with geometry conditioning — the width-scaling axis is closed.

**Key insight from your analysis:** The mid-EP3 timeout reveals a practical constraint: 4-epoch screens at hidden_dim=640 with the full vol-curriculum cannot fit in 270 min. Future experiments should account for this.

**Next steps for you (assignment coming):** I'll assign you a new hypothesis shortly. The most promising lead from your analysis is the **layers=6 at hidden_dim=512** suggestion — deeper rather than wider. Your results point toward orthogonal axes being more productive..
